posts/ebpf-and-kernel-modules-whats-the-difference.md

```diff
@@ -0,0 +1,115 @@
+
+Some network interface cards (NIC) support running an eBPF program type called [XDP](https://www.tigera.io/learn/guides/ebpf/ebpf-xdp/) that can process packets as soon as the NIC receives the packet and before the packet reaches the kernel. To do this with a linux kernel module, the NIC vendor would need to provide an SDK and would require the programmer to have a low-level knowledge of the card.
+
+With all this being said, the scope of eBPF might expand in the future. In the book _Learning eBPF_, the author says that work is currently being done to add hardware device support to eBPF.
+
+# Runtime
+
+The following is a simplified explanation.
+
+When kernel modules are loaded into the kernel, the kernel allocates memory for that program to execute, gives it all the privileges kernel code would get, and keeps a list tracking all kernel module names and locations so that the kernel can initialize and cleanup modules on user requests. Kernel module code runs directly on the CPU, like the kernel does.
+
+eBPF programs on the other hand will run as bytecode inside the eBPF VM, which is a sandboxed subsystem of the linux kernel. The bytecode is Just-in-Time (JIT) compiled into machine code during execution. This is similar to how Java bytecode runs on the JVM. The eBPF Linux subsystem will also keep track of running programs. 
+
+
+# Installing Applications
+
+When installing existing eBPF or Kernel Module programs, there's a good chance it will be installed just as any other Linux program is: via a package manager, a distributed binary, or an install script.
+
+Cilium ships a client CLI binary that is used to install the Cilium eBPF components onto kubernetes clusters. Falco is shipped as a package for both the eBPF and kernel module drivers. Wireguard is shipped as a package. Some kernel modules are shipped in Linux distributions and then can be enabled on install.
+
+Installing from source is a bit different than other programs.
+
+In eBPF's case, eBPF development packages are needed for compilation. Once the eBPF program is compiled, it will most likely be loaded into the eBPF kernel subsystem with the help of a user-space application that loads and manages the program. The user-space code will also likely be distributed with the eBPF code.
+
+In Kernel Module's case, kernel module development packages are needed for compilation. The build system is typically written in a Makefile. Once the binary is compiled, it will need to be loaded with kernel module management command `insmod`.
+
+
+# Supported Programming Languages
+eBPF programs can be written in any language that has a compiler that can output eBPF bytecode. Currently, and as far as I know, those languages are C, Rust, [BPFTrace](https://github.com/iovisor/bpftrace) and [P4](https://github.com/p4lang/p4c/tree/main).
+
+Kernel modules can be written in C or Rust.
```

Comment:
Remove spacing above and below and just make this whole section one paragraph.  Alternatively, add more information and keep it separate. The short single sentence paragraphs with a line in between are a bit jarring to the "flow".